python/basix/ufl_wrapper.py

```diff
@@ -1216,8 +1210,178 @@ def _create_enriched_element(elements: _typing.List[_FiniteElementBase]) -> _Fin
+                            tabulations.append(new_table)
+                        else:
+                            tabulations.append(e.tabulate(0, mapped_pts)[0].reshape((npts, -1, e.value_size)))
+
+                    ortho_tab = _basix.tabulate_polynomials(_basix.PolynomialType.legendre, ct, hd, mapped_pts)
+                    blocked_ortho_tab = _numpy.zeros((ortho_tab.shape[0] * vsize, npts * vsize))
+                    for v in range(vsize):
+                        blocked_ortho_tab[ortho_tab.shape[0] * v: ortho_tab.shape[0] * (v + 1),
+                                          v * npts: (v + 1) * npts] = ortho_tab
+                    blocked_ortho_tab = (wcoeffs @ blocked_ortho_tab).reshape((npoly, npts, vsize))
+
+                    orthogonal_data: _typing.List[_nda_f64] = []
+                    for table, e in zip(tabulations, elements):
+                        for p in range(e.dim):
+                            coeffs = [sum(table[i, p, v] * blocked_ortho_tab[q, i, v]
+                                      for i in range(npts) for v in range(vsize)) for q in range(npoly)]
+                            for o in orthogonal_data:
+                                coeffs -= _numpy.dot(o, coeffs) * o
+                            if not _numpy.isclose(sum(abs(i) for i in coeffs), 0):
+                                coeffs /= _numpy.dot(coeffs, coeffs) ** 0.5
+                                orthogonal_data.append(coeffs)
+                    orthogonal = _numpy.array(orthogonal_data)
+
+                    entity_ortho_tab = (orthogonal @ blocked_ortho_tab.reshape(npoly, -1)).reshape((-1, npts, vsize))
+
+                    rows = _numpy.zeros(entity_ortho_tab.shape)
+                    row_n = 0
+
+                    entities = [(d, i)] + [(d2, entity) for d2 in range(d) for entity in connectivity[d][i][d2]]
+                    for d2, entity in entities:
+                        for e, tab in zip(elements, tabulations):
+                            for dof in e.entity_dofs[d2][entity]:
+                                rows[row_n, :, :] = tab[:, dof, :]
+                                row_n += 1
+                    assert row_n == rows.shape[0]
+
+                    matrix = _numpy.array([[sum(sum(i) for i in f * g) for g in entity_ortho_tab] for f in rows])
+
+                    M_mat = _numpy.zeros((ndofs, vsize, npts, deriv_dim))
+
+                    for dof in range(ndofs):
+                        rhs = _numpy.array([1.0 if i == dof else 0.0 for i in range(matrix.shape[1])])
+                        coeffs = _numpy.linalg.solve(matrix, rhs)
+                        for v in range(vsize):
+                            M_mat[dof, v, :, 0] = coeffs @ entity_ortho_tab[:, :, v]
+
+                    M_entry.append(M_mat)
+
+                x.append(x_entry)
+                M.append(M_entry)
+
+        for i in range(tdim + 1, 4):
+            x.append([])
+            M.append([])
```

Comment:
Some documentation of what is going on here would be beneficial for the uneducated people (i.e. me)